### ISSUE TITLE
Incorrect instance type Parameter name: obj

### ISSUE THREAD

**Issue by sergey-tihon** [fsprojects/SwaggerProvider#101]
### Description

Reported by @tencek on gitter

> Hi, I updated from SwaggerProvider from 0.8.1 to 0.10.0-alpha9 to start using the *Async API. I'm getting an error I do not understand: error FS3033: The type provider 'SwaggerProvider.SwaggerTypeProvider' reported an error: Incorrect instance type Parameter name: obj I see the same error even with the petstore example from SwaggerProvider homepage. Here - they explain a possible reason but I can't figure out any action for my case. Any hints?

### Repro steps

Please provide the steps required to reproduce the problem

1. Restore the latest version from NuGet

2. Try sample from site

```fsharp
#load @"packages/SwaggerProvider/SwaggerProvider.fsx"
open SwaggerProvider

let [<Literal>]schema = "http://petstore.swagger.io/v2/swagger.json"
type PetStore = SwaggerProvider<schema, PreferAsync=true> // Provided Types
let petStore = PetStore()

``` 

### Expected behavior

Works =)

### Actual behavior

<img width="730" alt="2018-02-07_2253" src="https://user-images.githubusercontent.com/1197905/35938323-c9a37802-0c59-11e8-8bc5-13158fb939a9.png">

### Known workarounds

No

@baronfel do you have ideas? Looks like incorrect coercion?

**Comment by baronfel** [comment]
It does look like that, but that's a very odd situation because we have that very same code in the test suite:
```fsharp
open SwaggerProvider
open Expecto
open System

type PetStore = SwaggerProvider<"http://petstore.swagger.io/v2/swagger.json", PreferAsync = true>
type PetStoreNullable = SwaggerProvider<"http://petstore.swagger.io/v2/swagger.json", ProvideNullable = true>
let store = PetStore.Client()
```

**Comment by sergey-tihon** [comment]
Yes... the only difference is that it may load difference dependent assemblies ...

**Comment by baronfel** [comment]
So you think instead of using `typedefof<Async<_>>` and/or `typedefof<Task<_>>` I need to get the runtime-referenced assemblies and find those types in the runtime assemblies?

**Comment by sergey-tihon** [comment]
I hope no... in my understanding TP SDK should do it for us (and warn when it cannot)!
I think it may be an old & new version of json/yaml library that works differently ...

**Comment by baronfel** [comment]
I dunno if it is those dependencies or if it's the async/task stuff, but I did manage to do a version that looks for FSharp.Core and mscorlib form the explicit set of runtime resolved assemblies that the Provider Context gives to us. All the tests are passing, let me send a PR just in case.

**Comment by baronfel** [comment]
@sergey-tihon you pushed a beta with updates that isn't quite working for me on a new mono project :( I get the same instanciation error on 0.10.0-alpha9 and 0.10.0-beta1.

**Comment by baronfel** [comment]
Interesting: On mono 5.8, with your beta01, here's what I get when compiling a sample project against net461 with just the sample code above:
```
  /Users/chet/code/projects/swagger/Program.fs(6,17): error FS3033: The type provider 'SwaggerProvider.SwaggerTypeProvider' reported an error: Could not resolve field token 0x040000aa, due to: Could not load type of field '<StartupCode$SwaggerProvider-Runtime>.$SwaggerParser:deserializer@131' (1) due to: Could not load file or assembly 'YamlDotNet, Version=4.3.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. assembly:YamlDotNet, Version=4.3.0.0, Culture=neutral, PublicKeyToken=null type:<unknown type> member:<none> assembly:/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta01/lib/net45/SwaggerProvider.Runtime.dll type:$SwaggerParser member:<none> [/Users/chet/code/projects/swagger/swagger.fsproj]
```

**Comment by baronfel** [comment]
I added logging of the runtime assemblies to my local build of the provider and here's what I get at the moment:
```
runtime assemblies: [|"/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/mscorlib.dll";
  "/Users/chet/.nuget/packages/fsharp.core/4.2.3/lib/net45/FSharp.Core.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Core.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Data.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Drawing.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Numerics.dll";
  "/Users/chet/.nuget/packages/system.valuetuple/4.4.0/ref/net461/System.ValueTuple.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Xml.dll";
  "/usr/local/share/dotnet/sdk/NuGetFallbackFolder/newtonsoft.json/10.0.1/lib/net45/Newtonsoft.Json.dll";
  "/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.DesignTime.dll";
  "/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll";
  "/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.Runtime.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.IO.Compression.FileSystem.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Runtime.Serialization.dll";
  "/Library/Frameworks/Mono.framework/Versions/Current/Commands/../lib/mono/4.5/System.Xml.Linq.dll";
  "/Users/chet/.nuget/packages/yamldotnet/4.2.3/lib/net35/YamlDotNet.dll"|]
Root /Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll
Pattern ../../../Newtonsoft.Json*/lib/net45;../../../../newtonsoft.json*/**/lib/net45;../../../YamlDotNet*/lib/net35;../../../../yamldotnet*/**/lib/net35
Probing locations: ["/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll"]
Assembly not found!
Root /Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll
Pattern ../../../Newtonsoft.Json*/lib/net45;../../../../newtonsoft.json*/**/lib/net45;../../../YamlDotNet*/lib/net35;../../../../yamldotnet*/**/lib/net35
Probing locations: ["/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll"]
Assembly not found!
Root /Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll
Pattern ../../../Newtonsoft.Json*/lib/net45;../../../../newtonsoft.json*/**/lib/net45;../../../YamlDotNet*/lib/net35;../../../../yamldotnet*/**/lib/net35
Probing locations: ["/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.dll"]
Assembly not found!
```
This goes hand in hand with the following errors:
```
/Users/chet/code/projects/swagger/Program.fs(6,17): error FS3033: 
  The type provider 'SwaggerProvider.SwaggerTypeProvider' reported an error: 
    Could not resolve field token 0x040000aa, due to: 
    Could not load type of field '<StartupCode$SwaggerProvider-Runtime>.$SwaggerParser:deserializer@131' (1) due to: 
    Could not load file or assembly 'YamlDotNet, Version=4.2.4.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. assembly:YamlDotNet, Version=4.2.4.0, Culture=neutral, PublicKeyToken=null type:<unknown type> member:<none> assembly:/Users/chet/.nuget/packages/swaggerprovider/0.10.0-beta2/lib/net45/SwaggerProvider.Runtime.dll type:$SwaggerParser member:<none> [/Users/chet/code/projects/swagger/swagger.fsproj]
```

**Comment by sergey-tihon** [comment]
With latest NuGet it should resolve `YamlDotNet, Version=4.3.0.0` and `newtonsoft.json/10.0.3`
```
RESTRICTION: == net461
NUGET
  remote: https://www.nuget.org/api/v2
    FSharp.Core (4.3.3)
    Newtonsoft.Json (10.0.3)
    SwaggerProvider (0.10.0-beta01)
      FSharp.Core (>= 4.0)
      Newtonsoft.Json (>= 10.0 < 11.0)
      YamlDotNet (>= 4.3 < 4.4)
    YamlDotNet (4.3)
```

with latest changes in `dev` branch, I see that on my WinVM VS can resolve all dependencies and even can correctly work a couple of seconds before `Parameter name: obj `. This is weird ...

![vs_swagger_low](https://user-images.githubusercontent.com/1197905/35962377-2f69a608-0cc2-11e8-8d4e-1f00fdd991ad.gif)

**Comment by tencek** [comment]
@sergey-tihon - that's exactly what I'm observing. It seems to be perfectly OK until I try to call some of the generated function list.

**Comment by baronfel** [comment]
This is still super weird. I did some tinkering around the `typedefof` stuff, to use the `ProvidedTypeBuilder.MakeGenericInstance` instead, but that didn't fix the problem. On my mono 5.8 machine I still get assembly resolution errors trying to find YamlDotNet. @sergey-tihon  this is some WIP stuff that I put on top of your package update changes.

**Comment by baronfel** [comment]
May be worth raising a flag over at the TPSDK repo for advice?

**Comment by sergey-tihon** [comment]
Let's first fix issue with `YamlDotNet` on your machine, and repro the same behavior as mine
Here is repo with all dependencies (created on Mono) - https://github.com/sergey-tihon/swagger-issue101

Here is VS Code behavior on macOS
![sw_vs_code](https://user-images.githubusercontent.com/1197905/36058838-0e8fbe9c-0e3c-11e8-90b1-dcc3b98e7d04.gif)

**Comment by baronfel** [comment]
I can reproduce the odd editor behavior with the sample project, but the project compiles and runs fine.  I altered the code to retrieve pet 7316746567143381226 (did some testing via the API to find a known good pet id) and print it out and I get the expected Pet data:

For
```fsharp
[<EntryPoint>]
let main argv =
    let customers = petStore.GetPetById(7316746567143381226L) |> Async.RunSynchronously
    printfn "%O" customers
    0
```

I get 
```shell
{Id=Some(7316746567143381226); Category=Some({Id=Some(0); Name=Some(string)}); Name="Arka"; PhotoUrls=["string"]; Tags=Some(Program+PetStore+Tag[]); Status=Some(available)}
```

**Comment by baronfel** [comment]
I'm interested to see if the `ProvidedTypeBuilder` change has any effect on the test repo you made.

**Comment by sergey-tihon** [comment]
I pushed all changes in `SwaggerProvider 0.10.0-beta02` and updated test repo but looks like it behaves the same at least in VS Code on masOS.

If it compiles and runs without issues, but we see strange behavior in IDE (VS & Ionide on macOS and win) ... it looks like the bug is in TPSDK or IDE (I mean FSCS that run TPSDK)...

**Comment by sergey-tihon** [comment]
This issues fixed by #104 and released as [v0.10.0-beta03](https://www.nuget.org/packages/SwaggerProvider/0.10.0-beta03)

@tencek please try the latest prerelease.

**Comment by tencek** [comment]
Thank you, I will give it a try.
